### PR TITLE
feature/SDK-10

### PR DESCRIPTION
**Pull request by maikel-maas**

### CODE REVIEW

File: packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+        privateKeyHex: keyPair.privateKeyHex,
+        payloadVc: credential,
+      },
+      context,
+    )
+
+    const vpJwt = await createVPJwt(
+      {
+        keyPair,
+        did: this.configuration.issuer.toString(),
```
Comment (nklomp) [inline packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts:66]: This should be an argument to the method, not something configured

File: packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+  readonly methods: IEBSIAuthorizationClient = {
+    getOIDProviderMetadata: this.getOIDProviderMetadata.bind(this),
+    getOIDProviderJwks: this.getOIDProviderJwks.bind(this),
+    getPresentationDefinition: this.getPresentationDefinition.bind(this),
+    getAccessToken: this.getAccessToken.bind(this),
+    initiateSIOPDidAuthRequest: this.initiateSIOPDidAuthRequest.bind(this),
+    createSIOPSession: this.createSIOPSession.bind(this),
+    createOAuth2Session: this.createOAuth2Session.bind(this),
+  }
+  private configuration: GetOIDProviderMetadataSuccessResponse
```
Comment (nklomp) [inline packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts:37]: This is a very confusing name for something which is of type SuccessResponse and apparently OID metadata

File: packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+
+  //TODO Pass the configuration manually or call init
+  constructor(readonly baseUrl: string = 'https://api-pilot.ebsi.eu/authorisation/v4/') {
+    // TODO call configuration endpoint
```
Comment (nklomp) [inline packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts:41]: Not from constructor. Rather would see a lazy method being called as a first action of any method that needs the metdata

File: packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+      },
+      context,
+    )
+
+    let descriptorMap: Descriptor[] = []
+
+    if (definitionId === ScopeByDefinition.didr_invite_presentation) {
+      descriptorMap = [
+        {
+          id: 'didr_invite_credential',
```
Comment (nklomp) [inline packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts:79]: The only difference in these descriptors is the id in 2 places. Suggest to create a function and pass in the id argument

File: packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+    })
+    if ('status' in tokenResponse) {
+      throw new Error(JSON.stringify(tokenResponse))
+    }
+    return tokenResponse
+  }
+
+  private async getOIDProviderMetadata(): Promise<GetOIDProviderMetadataResponse> {
+    return await (
+      await fetch(`${this.baseUrl}/.well-known/openid-configuration`, {
```
Comment (nklomp) [inline packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts:137]: We should have generic fetch support, which does proper error handling. Either in ssi-types, sdk-core or express-support from the top of my head

File: packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+    throw new Error('Verifiable Credential not defined')
+  }
+
+  const privateKeyJWT = await exportJWK(u8a.fromString(keyPair.privateKeyHex, 'hex'))
+  const publicKeyJWT = await exportJWK(u8a.fromString(keyPair.publicKeyHex, 'hex'))
+
+  let verifiableCredential: string[]
+
+  if (vc === 'empty') {
+    verifiableCredential = []
```
Comment (nklomp) [inline packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts:259]: How would no VCs result in a valid VP where the definitions already made clear it expects 1 VC?

File: packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+
+  if (!keyPair || !keyPair.privateKeyHex || !keyPair.publicKeyHex) {
+    throw new Error(`Private key is required`)
+  }
+
+  if (!args.vc) {
+    throw new Error('Verifiable Credential not defined')
+  }
+
+  const privateKeyJWT = await exportJWK(u8a.fromString(keyPair.privateKeyHex, 'hex'))
```
Comment (nklomp) [inline packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts:253]: Never ever handle private keys outside the KMS

File: packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts [hunk 8]
```diff
@@ -0,0 +1,363 @@
...
+    await signingKey.export({
+      type: 'JsonWebKey2020',
+      privateKey: true,
+    }),
+    {
+      detached: false,
+      header: jwtHeader,
+    },
+  )
+  const vpSigner = vpKey.signer()
```
Comment (nklomp) [inline packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts:319]: Don;t use external libs for this. The agent is perfectly capable of creating VCs and VPs of course!

File: packages/ebsi-authorization-client/src/types/IEBSIAuthorizationClient.ts [hunk 9]
```diff
@@ -0,0 +1,310 @@
+import { IAgentContext, ICredentialPlugin, IDIDManager, IKeyManager, IPluginMethodMap } from '@veramo/core'
+import { Format, PresentationDefinitionV2, PresentationSubmission } from '@sphereon/pex-models'
+
+/**
+ * The OpenID scope
+ * @readonly
+ * @enum {string}
+ */
+export enum OpenIDScope {
+  didr_write = 'openid%20didr_write',
```
Comment (nklomp) [inline packages/ebsi-authorization-client/src/types/IEBSIAuthorizationClient.ts:10]: Hmzz, I don;t like that. The openId scope is litterally only "openid". Then EBSI is defining other scopes as well. Their eindpoints expect 2 scopes. Calling these OpenID scopes and then already combining them is not nice. Simply ensure the scope is a seperate enum and ensure the openid scope is always present when constructing the scope

File: packages/ebsi-authorization-client/src/types/IEBSIAuthorizationClient.ts [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+ * @property {string[]} [request_parameter_supported] Boolean value specifying whether the OP supports use of the request parameter, with true indicating support
+ * @property {string[]} [token_endpoint_auth_methods_supported] JSON array containing a list of client authentication methods supported by this token endpoint
+ * @property {{ authorization_endpoint: string[] }} [request_authentication_methods_supported] A JSON Object defining the client authentications supported for each endpoint
+ * @property {string[]} [vp_formats_supported] An object containing a list of key value pairs, where the key is a string identifying a credential format supported by the AS
+ * @property {(URL[] | string[])} [subject_syntax_types_supported]  A JSON array of strings representing URI scheme identifiers and optionally method names of supported Subject Syntax Types
+ * @property {string[]} [subject_trust_frameworks_supported] A JSON array of supported trust frameworks.
+ * @property {string[]} [id_token_types_supported] A JSON array of strings containing the list of ID Token types supported by the OP
+ */
+export interface GetOIDProviderMetadataSuccessResponse {
```
Comment (nklomp) [inline packages/ebsi-authorization-client/src/types/IEBSIAuthorizationClient.ts:68]: Why are we redefining what we have available in our SIOP libs?

File: packages/ebsi-authorization-client/src/types/IEBSIAuthorizationClient.ts [hunk 11]
```diff
@@ -0,0 +1,310 @@
...
+ * @typedef Key
+ * @type {object}
+ * @property {string} [kty]
+ * @property {string} [crv]
+ * @property {string} [alg]
+ * @property {string} [x]
+ * @property {string} [y]
+ * @property {string} [kid]
+ */
+export interface Key {
```
Comment (nklomp) [inline packages/ebsi-authorization-client/src/types/IEBSIAuthorizationClient.ts:100]: Why do we need to define this. Not like we are not using keys to elsewhere to begin with. Next to that this is actually a JWK it seems, so please call it a JWK then

File: packages/ebsi-authorization-client/src/types/IEBSIAuthorizationClient.ts [hunk 12]
```diff
@@ -0,0 +1,310 @@
...
+ * @type {object}
+ * @property {string} [ake1_enc_payload] Encrypted payload with user's public key
+ * @property {string} [ake1_nonce] Nonce used during the authentication process
+ * @property {string} [did] API DID
+ * @property {string} [kid] Trusted App KID
+ * @property {number} [iat] Issued at
+ * @property {number} [exp] Expires at
+ * @property {string} [iss] Issuer (Authorisation API)
+ */
+export interface Ake1SigPayload {
```
Comment (nklomp) [inline packages/ebsi-authorization-client/src/types/IEBSIAuthorizationClient.ts:231]: What is this doing here?

**Review by nklomp — CHANGES_REQUESTED** [review]
This will need a significant amount of changes. Never ever require private keys in methods, except for the KMS no agent plugin can access private keys

File: packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts (hunk 4 above)
Comment (maikel-maas) [reply packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts:79]: Extracted into a method: getDescriptorMap(...)

File: packages/ebsi-authorization-client/src/types/IEBSIAuthorizationClient.ts (hunk 9 above)
Comment (maikel-maas) [reply packages/ebsi-authorization-client/src/types/IEBSIAuthorizationClient.ts:10]: Changed the name of the enum to EBSIScope and added openid scope directly to the request

File: packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts (hunk 7 above)
Comment (maikel-maas) [reply packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts:234]: Searching for the key in kms using a kid now

File: packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts (hunk 6 above)
Comment (maikel-maas) [reply packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts:240]: Something inherited from the EBSI code base. Removed

File: packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts (hunk 8 above)
Comment (maikel-maas) [reply packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts:300]: Removed third-party libraries and signing JWTs with the agent now

File: packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts (hunk 1 above)
Comment (maikel-maas) [reply packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts:59]: Passing and argument to the method now

File: packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts (hunk 2 above)
Comment (maikel-maas) [reply packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts:32]: Removed based on the comment below

File: packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts [hunk 13]
```diff
@@ -0,0 +1,363 @@
...
+  GetPresentationDefinitionArgs,
+  GetPresentationDefinitionResponse,
+  IEBSIAuthorizationClient,
+  InitiateSIOPDidAuthRequestArgs,
+  InitiateSIOPDidAuthRequestResponse,
+} from '../types/IEBSIAuthorizationClient'
+import { uuid } from 'uuidv4'
+import { Descriptor } from '@sphereon/pex-models'
+import { JWTHeader, JWTPayload } from 'did-jwt'
+import { exportJWK, importJWK, JWK, SignJWT } from 'jose'
```
[the remark this answers is not in the document]
Comment (maikel-maas) [reply packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts]: Removed jose

File: packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts [hunk 14]
```diff
@@ -0,0 +1,363 @@
...
+import * as u8a from 'uint8arrays'
+import { JsonWebKey, JsonWebKey2020 } from '@transmute/json-web-signature'
```
[the remark this answers is not in the document]
Comment (maikel-maas) [reply packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts]: Removed @transmute/json-web-signature

File: packages/ebsi-authorization-client/src/types/IEBSIAuthorizationClient.ts (hunk 11 above)
Comment (maikel-maas) [reply packages/ebsi-authorization-client/src/types/IEBSIAuthorizationClient.ts:83]: Importing JWK from @sphereon/did-auth-siop

File: packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts (hunk 5 above)
Comment (maikel-maas) [reply packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts:112]: There only one of the 3 libraries that has cross-fetch as a dependency: ssi-sdk-core. It's used specifically to download images. What do I miss?

File: packages/ebsi-authorization-client/src/types/IEBSIAuthorizationClient.ts (hunk 12 above)
Comment (maikel-maas) [reply packages/ebsi-authorization-client/src/types/IEBSIAuthorizationClient.ts:196]: This is the response from the EBSI RP

File: packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts [hunk 15]
```diff
@@ -0,0 +1,387 @@
...
+  IEBSIAuthorizationClient,
+  InitiateSIOPDidAuthRequestArgs,
+  InitiateSIOPDidAuthRequestResponse,
+} from '../types/IEBSIAuthorizationClient'
+import { uuid } from 'uuidv4'
+
+export class EBSIAuthorizationClient implements IAgentPlugin {
+  readonly schema = schema.IEBSIAuthorizationClient
+  readonly methods: IEBSIAuthorizationClient = {
+    getOIDProviderMetadata: this.getOIDProviderMetadata.bind(this),
```
Comment (nklomp) [inline packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts:24]: Please prefix them all with 'ebsi', like we do in other agent methods. Remember that these methods will all become available if you include this plugin in an agent. Then a method called "getAccessToken" has no relationship for a developer anymore to EBSI. Hence why we prefix plugin methods, so they make sense to an external developer and are also grouped together

File: packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts [hunk 16]
```diff
@@ -0,0 +1,387 @@
...
+    return tokenResponse
+  }
+
+  async siop(args: { scope: EBSIScope; callbackUrl: string; did: string }, context: IRequiredContext) {
+    const { scope, callbackUrl, did } = args
+
+    const authRequest = await this.initiateSIOPDidAuthRequest({ scope })
+
+    // TODO add proper error handling
+    if (!(typeof authRequest === 'string')) {
```
Comment (nklomp) [inline packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts:93]: Why not !==?

File: packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts (hunk 15 above)
Comment (maikel-maas) [reply packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts:24]: Prefixed the methods with ebsiAuth

File: packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts [hunk 17]
```diff
@@ -0,0 +1,418 @@
...
+      id: uuid(),
+      definition_id: id,
+      descriptor_map: descriptorMap,
+    }
+
+    const tokenResponse = await this.getAccessToken({
+      grant_type: 'vp_token',
+      vp_token: vpJwt,
+      presentation_submission: presentationSubmission,
+      scope: EBSIScope.didr_invite,
```
Comment (jcmelati) [inline packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts:207]: Should this `scope` come from args instead of being hardcoded?

Comment (maikel-maas) [reply packages/ebsi-authorization-client/src/agent/EBSIAuthorizationClient.ts]: Retrieving the values from args now

File: packages/ebsi-authorization-client/src/types/IEBSIAuthorizationClient.ts (hunk 12 above)
Comment (maikel-maas) [reply packages/ebsi-authorization-client/src/types/IEBSIAuthorizationClient.ts:219]: Not needed anymore

**Pull request merged**